homeassistant/components/fibaro.py

```diff
@@ -0,0 +1,360 @@
+        controller.enable_state_handler()
+        return True
+    return False
+
+
+class FibaroDevice(Entity):
+    """Representation of a Fibaro device entity."""
+
+    def __init__(self, fibaro_device, controller):
+        """Initialize the device."""
+        self.fibaro_device = fibaro_device
+        self.controller = controller
+        self._name = fibaro_device.friendly_name
+        self.ha_id = fibaro_device.ha_id
+        self.controller.register(fibaro_device.id, self._update_callback)
+
+    def _update_callback(self):
+        """Update the state."""
+        self.update()
+        self.schedule_update_ha_state(False)
+
+    def get_level(self):
+        """Get the level of Fibaro device."""
+        if 'value' in self.fibaro_device.properties:
+            return self.fibaro_device.properties.value
+        return None
+
+    def get_level2(self):
+        """Get the tilt level of Fibaro device."""
+        if 'value2' in self.fibaro_device.properties:
+            return self.fibaro_device.properties.value2
+        return None
+
+    def dont_know_message(self, action):
+        """Make a warning in case we don't know how to perform an action."""
+        _LOGGER.warning("Not sure how to setValue: %s "
+                        "(available actions: %s)", str(self.ha_id),
+                        str(self.fibaro_device.actions))
+
+    def set_level(self, level):
+        """Set the level of Fibaro device."""
+        self.action("setValue", level)
+        if 'value' in self.fibaro_device.properties:
+            self.fibaro_device.properties.value = level
+        if 'brightness' in self.fibaro_device.properties:
+            self.fibaro_device.properties.brightness = level
+
+    def set_color(self, red, green, blue, white):
+        """Set the tilt level of Fibaro device."""
+        color_str = "{},{},{},{}".format(int(red), int(green),
+                                         int(blue), int(white))
+        self.fibaro_device.properties.color = color_str
+        self.action("setColor", str(int(red)), str(int(green)),
+                    str(int(blue)), str(int(white)))
+
+    def action(self, cmd, *args):
+        """Perform an action on the Fibaro HC."""
+        if cmd in self.fibaro_device.actions:
+            getattr(self.fibaro_device, cmd)(*args)
+            _LOGGER.debug("-> %s.%s%s called", str(self.ha_id),
+                          str(cmd), str(args))
+        else:
+            self.dont_know_message(cmd)
+
+    @property
+    def hidden(self) -> bool:
+        """Return True if the entity should be hidden from UIs."""
+        return self.fibaro_device.visible is False
+
+    @property
+    def current_power_w(self):
+        """Return the current power usage in W."""
+        if 'power' in self.fibaro_device.properties:
+            power = self.fibaro_device.properties.power
+            if power:
+                return convert(power, float, 0.0)
+        else:
+            return 0
```

Comment:
`return None` rather than zero?

Reply:
It feels like `current_power_w` should not be in the base class.

Reply:
Agreed that code copying should be avoided. I was thinking that you could have a `FibaroPowerDevice` and not have to decide capabilities dynamically.

Anyway, let's just keep it as is for now.